docs/dev/framework/migrations.md

```diff
@@ -0,0 +1,117 @@
+---
+title: "Migrations"
+description: "Database migrations and general purpose migration scripts."
+---
+
+{{% notice note %}}
+This covers the documentation on how to create migrations in Contao **4.9**
+and up. In previous Contao versions, migrations were written in `runonce.php` files that got deleted after execution.
+{{% /notice %}}
+
+Updating Contao, extensions or the application itself sometimes requires to migrate data to be compatible with the new version(s). For this purpose Contao has a migration framework that lets you write migration services that are integrated in the update process.
+
+The migrations get executed via the install tool database update or with the [`contao:migrate` command][commands].
+
+
+## Definition
+
+To add a new migration, create a service that implements the interface `Contao\CoreBundle\Migration\MigrationInterface` and add the tag `contao.migration` (if autoconfiguration is enabled, this happens automatically):
+
+```yaml
+# config/services.yaml
+services:
+    App\Migration\MyMigration:
+        tags:
+            - { name: contao.migration, priority: 0 }
+```
+
+## MigrationInterface
+
+The migration interface specifies three methods that need to be implemented:
+
+* __getName()__<br>
+  A name that describes what the migration does. This text is shown to users when they are asked if they want to execute the migration.
+
+* __shouldRun()__<br>
+  This method checks if all prerequisites that are needed for the migration to run are met and if it actually needs to run. This method should be written very defensively because the application might be in an unexpected state when the method gets called, e.g. the database could be completely empty.
+
+* __run()__<br>
+  As the name suggests, that is where the real magic happens. If `shouldRun()` returned `true`, this method will be called and should do the actual migration.<br>
+  It returns a `MigrationResult` object that can hold more information about what happened during the execution and if the migration was successful or not.<br>
+  If something goes unexpectedly wrong here and you want to abort the migration process completeley you should throw an exception here.
+
+
+## Example
+
+Lets say we have a database table `tl_customers` with a `firstName` and `lastName` column that we combined to a `name` column in the new version:
+
+```php
+// src/Migration/CombineCustomerName.php
+namespace App\Migration;
+
+use Contao\CoreBundle\Migration\AbstractMigration;
+use Contao\CoreBundle\Migration\MigrationResult;
+use Doctrine\DBAL\Connection;
+
+class CombineCustomerName extends AbstractMigration
```

Comment:
Minor nitpick: shouldn't the class name be suffixed with `Migration` here? I actually thought it was a general coding style rule within Symfony to suffix the class name with its namespace, to make it more clear what the class does. But that's actually not the case.

Reply:
Done in 22a135ce17b849f2b48266d3f0087ef3679cce6d